File: C++/Divide-Two-Integers.cpp
class Solution {
public:
    /**
     * @param dividend the dividend
     * @param divisor the divisor
     * @return the result
     */
    int divide(int dividend, int divisor) {
        // Write your code here
        int symbol = 1;
        long long dvd = dividend, dvr = divisor, res = 0;
        if (!((dividend >= 0 && divisor >= 0) ||
                (dividend < 0 && divisor < 0))) symbol = -1;
        if (dvd < 0) dvd = -dvd;
        if (dvr <  0) dvr = -dvr;
        while (dvd >= dvr) {
            long long cnt = dvr, num = 1;
            while ((cnt << 1) <= dvd) {
                cnt = cnt << 1;
                num = num << 1;
            }
            dvd -= cnt;
            res += num;
        }
        res = res * symbol;
        if (res > INT_MAX || res < INT_MIN) return INT_MAX;
        return res;
    }
};

File: Python/Divide-Two-Integers.py
class Solution:
    # @return an integer
    def divide(self, dividend, divisor):
        symbol = 1
        if dividend < 0 and divisor < 0:
            dividend = - dividend
            divisor = - divisor
        elif dividend < 0:
            dividend = - dividend
            symbol = -1
        elif divisor < 0:
            divisor = - divisor
            symbol = -1
        cnt = divisor
        ans = 1
        res = 0
        while dividend > 0:
            if dividend < divisor:
                break
            while cnt < dividend and cnt + cnt < dividend:
                cnt += cnt
                ans += ans
            dividend -= cnt
            res += ans
            ans = 1
            cnt =divisor
        if res * symbol > 2147483647:
            return 2147483647
        else:
            return res * symbol